design/secrets_provider_milestone1_2_design.md

```diff
@@ -0,0 +1,380 @@
+
+This is problematic for us at this stage because now the Secrets Provider is sitting separately and will need to perform a batch retrieval of all the secrets in the namespace. In other words, if the Secrets Provider fails to fetch one Conjur secret, the whole request will fail and applications will not get their secrets. To avoid this, we will update the code so that if a failure takes place, it will be noted in the logs and a new request will be made, skipping over the Conjur secret that caused the error.
+
+```pseudocode
+for each k8s-secret
+  variable-ids = k8s-secret.parseConjurMap()
+  GET /secrets{?variable-ids}
+```
+
+
+
+#### Fetching Relevant K8s Secrets
+
+With the decoupling of Secrets Provider and application pod, we can know what types of K8s Secrets the pod requires by either of the following ways: 
+
+|      | Fetch/Update K8s Secret                                      | Elaboration                                                  | Pros                                                         | Cons                                        |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------- |
+| 1    | Read all K8s Secrets that are marked with labels and update them with Conjur secrets |                                                              | - Scalable <br /><br />- Less intervention on our part as a customer can define which types of secrets we should parse<br /><br />- K8s native, as-is solution | - Not generalized so we are still K8s bound |
+| 2    | Get K8s secrets name by configuration in Secrets Provider    | Define K8S_SECRETS in `cyberark-secrets-provider` manifest<br />` name: K8S_SECRET` | - No code changes<br /><br />- Backwards compatible          | - Repetitive work<br /><br />- Not scalable |
+
+To know which K8s Secrets the apps in the namespace need we chose the first solution, Read all K8s Secrets that are marked with labels. It would be the most K8s native and avoids the repetitive work of having to list all the K8s Secrets each app needs.
+
+It is important to note however, that we are responsible for providing a solution for updating the K8s secrets and *not* how those Pods will get those secrets.
+
+##### **Customer Experience**
+
+K8s offers the option of adding labels on K8s secrets. For example, a customer would add a label on a K8s Secret like so:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: db-secret
+  labels:
+    conjur-secret: env-prod # label name must be 'conjur-secret'; value is supplied by the customer
```

Comment:
1. let's get an approval for the label name.
2. what does the value represent? what is `env-prod`?

Reply:
I updated it to make the label key configurable because are parsing the keys

Reply:
I updated this to be so an admin can configure the label key. 
From what I have seen in K8s, the key is the category and the value is the state.
Ex:
`environment` : `prod`